starlite/datastructures/url.py

```diff
@@ -0,0 +1,212 @@
+from typing import TYPE_CHECKING, Any, NamedTuple, Optional, Union
+from urllib.parse import SplitResult, urlencode, urlsplit, urlunsplit
+
+from multidict import MultiDict
+
+from starlite.datastructures import Headers
+from starlite.datastructures.multi_dicts import QueryMultiDict
+
+if TYPE_CHECKING:
+    from starlite.types import Scope
+
+
+class Address(NamedTuple):
+    """Just a network address."""
+
+    host: str
+    port: int
+
+
+def make_absolute_url(path: Union[str, "URL"], base: Union[str, "URL"]) -> str:
+    """Create an absolute URL.
+
+    Args:
+        path: URL path to make absolute
+        base: URL to use as a base
+
+    Returns:
+        A string representing the new, absolute URL
+    """
+    if isinstance(base, str):
+        base = URL(base)
+    netloc = base.netloc
+    path = base.path.rstrip("/") + str(path)
+    return str(URL.from_components(scheme=base.scheme, netloc=netloc, path=path))
+
+
+class URL:
+    __slots__ = (
+        "scheme",
+        "netloc",
+        "path",
+        "fragment",
+        "query",
+        "username",
+        "password",
+        "port",
+        "hostname",
+        "_query_params",
+        "_url",
+    )
+
+    scheme: str
+    """URL scheme"""
+    netloc: str
+    """Network location"""
+    path: str
+    """Hierarchical path"""
+    fragment: str
+    """Fragment component"""
+    query: str
+    """Query string"""
+    username: Optional[str]
+    """Username if specified"""
+    password: Optional[str]
+    """Password if specified"""
+    port: Optional[int]
+    """Port if specified"""
+    hostname: Optional[str]
+    """Hostname if specified"""
+
+    def __init__(self, url: Union[str, SplitResult]) -> None:
+        """Representation and modification utilities of a URL.
+
+        Args:
+            url: URL, either as a string or a [SplitResult][urllib.parse.SplitResult] as returned by [urlsplit][urllib.parse.urlsplit]
+        """
+        if isinstance(url, str):
+            result = urlsplit(url)
+            self._url = url
+        else:
+            result = url
+            self._url = urlunsplit(url)
+
+        self.scheme = result.scheme
+        self.netloc = result.netloc
+        self.path = result.path
+        self.fragment = result.fragment
+        self.query = result.query
+        self.username = result.username
+        self.password = result.password
+        self.port = result.port
+        self.hostname = result.hostname
+        self._query_params: Optional[QueryMultiDict] = None
+
+    @classmethod
+    def from_components(
+        cls,
+        scheme: Optional[str] = None,
```

Comment:
We have logic in place that uses `None` for some of these. Could be cleaned up, but I wasn't sure if this could cause issue with user facing code, so it left it as-is.